app/src/main/java/com/feedhenry/securenativeandroidtemplate/domain/crypto/PreAndroidMSecureKeyStore.java

```diff
@@ -77,58 +79,57 @@ public void generateAESKey(String keyAlias) throws GeneralSecurityException, IOE
     }
 
     @Override
-    public void deleteKey(String keyAlias) throws GeneralSecurityException, IOException {
-        if (hasKeyAlias(keyAlias)) {
-            SharedPreferences.Editor editor = this.sharedPreferences.edit();
-            editor.remove(keyAlias);
-            editor.commit();
-        }
-    }
-
-    private byte[] generateSecretKey(int size) {
-        byte[] secretKey = new byte[size/BIT_PER_BYTE];
-        SecureRandom secureRandom = new SecureRandom();
-        secureRandom.nextBytes(secretKey);
-        return secretKey;
-    }
-
-    private void generateRSAKeyForAndroidM() throws GeneralSecurityException, IOException {
+    public void generatePrivateKeyPair(String keyAlias) throws GeneralSecurityException, IOException {
         //pre android-M, the keystore only support RSA key generation. So here we will generate a RSA keypair first, then generate the AES key.
         //we then encrypt the AES key using the generated RSA public key, and save it using the SharedPreferences
         Calendar start = Calendar.getInstance();
         Calendar end = Calendar.getInstance();
         end.add(Calendar.YEAR, 99);
         KeyPairGeneratorSpec generatorSpec = new KeyPairGeneratorSpec
                 .Builder(context)
-                .setAlias(RSA_KEY_ALIAS)
-                .setSubject(new X500Principal("CN=" + RSA_KEY_ALIAS))
+                .setAlias(keyAlias)
+                .setSubject(new X500Principal("CN=" + keyAlias))
                 .setSerialNumber(BigInteger.TEN)
                 .setStartDate(start.getTime())
                 .setEndDate(end.getTime())
                 .build();
         KeyPairGenerator generator = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_RSA, ANDROID_KEY_STORE);
         generator.initialize(generatorSpec);
         generator.generateKeyPair();
+    }
+
+    @Override
+    public boolean hasKeyPair(String keyAlias) throws GeneralSecurityException, IOException {
+        KeyStore ks = loadKeyStore();
+        return ks.containsAlias(keyAlias);
+    }
+
+    @Override
+    public void deleteKey(String keyAlias) throws GeneralSecurityException, IOException {
+        if (hasSecretKey(keyAlias)) {
+            SharedPreferences.Editor editor = this.sharedPreferences.edit();
+            editor.remove(keyAlias);
+            editor.commit();
+        } else if (hasKeyPair(keyAlias)) {
+            KeyStore ks = loadKeyStore();
+            ks.deleteEntry(keyAlias);
+        }
+    }
 
+    private byte[] generateSecretKey(int size) {
+        byte[] secretKey = new byte[size/BIT_PER_BYTE];
+        SecureRandom secureRandom = new SecureRandom();
```

Comment:
I think it might be better suited to use the [javax.crypto.KeyGenerator](https://docs.oracle.com/javase/7/docs/api/javax/crypto/KeyGenerator.html) here to generate the AES key.